File: top_interview_questions/1_easy/3_linked_list/remove_nth_node_from_end_of_list/solution.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

func removeNthFromEnd(head *ListNode, n int) *ListNode {
	length := getLength(head)

	if n == 1 && length == 1 {
		return nil
	}

	return removeNthElement(head, length-n)
}

func removeNthElement(head *ListNode, n int) *ListNode {
	node := head

	for i := 0; i < n-1; i++ {
		node = node.Next
	}

	if node.Next != nil {
		if n == 0 {
			node.Val = node.Next.Val
		}
		node.Next = node.Next.Next
	}

	return head
}

func getLength(head *ListNode) int {
	i := 0

	for head != nil {
		head = head.Next
		i++
	}

	return i
}

File: top_interview_questions/1_easy/3_linked_list/remove_nth_node_from_end_of_list/solution.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        length = self.getLength(head)

        if n == 1 and length == 1:
            return None

        return self.removeNthElement(head, length - n)

    def removeNthElement(self, head: ListNode, n: int) -> ListNode:
        node = head
        for _ in range(n - 1):
            node = node.next

        if node.next:
            if n == 0:
                node.val = node.next.val
            node.next = node.next.next

        return head

    def getLength(self, head: ListNode) -> int:
        if not head:
            return 0

        return self.getLength(head.next) + 1